### ISSUE TITLE
Example try expression in Typescript

### ISSUE THREAD

**Issue by Arlen22** [arthurfiorette/proposal-safe-assignment-operator#48]
I've updated the current Typescript to include this operator as a passthrough (as though it's implemented in Javascript) so we can play around with examples a bit. 

https://github.com/Arlen22/TypeScript - main branch

I basically just implemented it as an expression operator, based off void and await. 

To try it, clone the repo, run the following...

```
npm install
npm run pack
npm pack
```

then npm install the output file in an example project. If you're using vscode you can change the tsserver location to get intellisense as well.

**Comment by Arlen22** [comment]
The details are basically that when it encounters the try keyword, it first tries to parse it as a statement, then as an expression, unless it's already in an expression context, in which case it parses it as an expression (curly braces become objects). 

I'm not sure I like that last part. I think curly braces should be an error unless they are surrounded by parentheses. 

I haven't tested any unusual parts of Javascript yet, like using or for loops.

**Comment by Arlen22** [comment]
For actual implementation, I think this should be converted to a self-executing function that returns the appropriate values. It's pretty standard for various syntax types to be converted into helper functions until they are implemented in Javascript. 

I'll probably do that sometime.

**Comment by arthurfiorette** [comment]
Hey @Arlen22, could you test if it aligns with the new rewriting?

https://github.com/arthurfiorette/proposal-safe-assignment-operator/pull/53

**Comment by Arlen22** [comment]
@arthurfiorette well, technically the parsing would still work. You'd have to edit the Result type. But I copied the code for the await and yield keywords, so it's valid for all expressions, not just the assignment expression. If I was going to do it again for the new proposal, I'd do it as a modifier of the assignment operator.

**Comment by Arlen22** [comment]
@arthurfiorette, I've updated it to fix a few things and add working JavaScript. It should work now for actually compiled code. 

The compiler basically generates a custom inline function that changes based on whether yield and await are used in the wrapped expression. It's probably not perfect.

**Comment by Lookwe69** [comment]
One word, thanks for the proposal !

For my part, I implemented a version of the proposal with a function.
It handles both synchronous and asynchronous operations. It can be useful for testing the result object.

[Github](https://github.com/Lookwe69/try)
[NPM](https://www.npmjs.com/package/@lookwe/try)

**Comment by Arlen22** [comment]
@Lookwe69 Nice. You might want to take a look at https://github.com/arthurfiorette/proposal-try-operator/issues/69#issuecomment-2697864186 which handles the other scenarios the try operator may find itself in.

**Comment by ljharb** [comment]
Because it's iterable, and iterables aren't necessarily indexable.

**Comment by ljharb** [comment]
I'm not sure that's always true - react hooks don't have to return an array, as far as I know, just an iterable. Either way, I'm not sure why the language would ever be constrained by something React does?

**Comment by DScheglov** [comment]
Yes, React don't have to return a tuple. But it is a widely spread pattern. 
The language should account the currect practice, or am I wrong?

**Comment by ljharb** [comment]
"react hooks" isn't the current practice this proposal is targeting, in any way, as far as i understand it - it's about errors, not about react hooks (in which, errors throw an exception)

**Comment by DScheglov** [comment]
You see... This proposal is about tuples precisely because such a practice has become quite widespread by React.
This made it very easy to explain the benefits of tuples -- because many users already have relevant experience.

Now we are proposing a very similar approach, but at the same time, in a number of edge cases, developers will have a completely different experience. 

The React team could have easily invented the same kind of tuples -- ones that are only iterable but not arrays (it could be easily explained by "performance" improvments) -- but they chose a simpler and more straightforward way. 

We are inventing the `Result` type.

Actually, we might run into another issue: what about the `map` method? How should it behave? Because if we add the `Result` type to JavaScript, it's only a matter of time before `map` and `flatMap` are proposed as well.

So, the more I think about this approach, the more I believe we should simply return something like the following:

```js
Object.assign([ok, error, value], { ok, error, value });
```

Do we really need the Result type for the try operator?

**Comment by ljharb** [comment]
Either way, the pattern used in React hooks is to *exclusively* destructure the iterable return result of a hook. Nobody ever indexes into that array - React could return a Set instead of an array and nobody's code would break.

I'm not sure what would be wrong with a map or flatMap method being proposed?

**Comment by DScheglov** [comment]
The pattern is to return an array-based tuple and then destructure it. How this array goes from being created to being destructured can be a bit tricky, and in that process some code might break if the array gets replaced with set.

Actually, I’m ok with not supporting index access. We just need to mention this in the proposal and note that it’s not likely anyone will need this feature. If they do, the [...result][index] is a simple workaround.

About `map` and `flatMap`.

Result is both an array‑like structure and a classic sum type known as Either (in Rust this type is also called Result).
Right now, on the TypeScript side, especially for the "polyfill", we have a problem: we are forced to have map and flatMap behave like they do on arrays, rather than like they do on Either.

Also, `[Symbol.iterator]` is implemented for destructuring and doesn’t help with emulating do‑notation, the way libraries like Effect‑TS do.

```js
const result = Result.do(function* () {
  const r1 = yield* try doSomething();
  const r2 = yeild* try doSomethingElse();

 return { r1, r2 }
});
```

It is equivalent to:

```js
const result = (() => {
  const r1=try doSomething();
  if (!r1.ok) return r1;
 
  const r2 = try doSomethingElse();
  if (!r2.ok) return r2;  

  return Result.ok({ r1: r1.value, r2: r2.value });
})()
```
